### PR TITLE
Add filter hook to allow currently hardcoded 10 minute checkout draft stock reservation to be filtered

### LINKED ISSUE #45245: Checkout draft orders from blocks do not respect woocommerce_hold_stock_minutes (can it? or add filter to allow < or > 10 minutes)
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

In `/src/StoreApi/Routes/V1/Checkout.php` in the `create_or_update_draft_order` function, there is this line of code:

```
$reserve_stock = new ReserveStock();
$duration      = $request->get_method() === 'POST' ? (int) get_option( 'woocommerce_hold_stock_minutes', 60 ) : 10;
$reserve_stock->reserve_stock_for_order( $this->order, $duration );
```

This is forcing stock to be reserved for a hardcoded 10 minutes for checkout draft orders.

I suggest this should either:

1. Use woocommerce_hold_stock_minutes as this is a way to limit/disable stock reservation entirely, but checkout drafts ignore this
2. Have a filter hook so the duration of this can be modified or disabled where you do not wish for checkout drafts to reserve stock for 10 minutes or at all

 

### Expected behavior

Stock reserved for checkout drafts should use the woocommerce_hold_stock_minutes setting OR let developers amend the hardcoded 10 minutes via a filter hook.

### Actual behavior

Checkout drafts have a 10 minute hold period which cannot be changed (without any clarity it is doing this over the woocommerce_hold_stock_minutes setting).

### Steps to reproduce

On a checkout draft order see the order notes have reserved stock for 10 minutes.

### WordPress Environment

`
### WordPress Environment ###

WordPress address (URL): https://amusingsail.s3-tastewp.com
Site address (URL): https://amusingsail.s3-tastewp.com
WC Version: 8.6.1
REST API Version: ✔ 8.6.1
WC Blocks Version: ✔ 11.8.0-dev
Action Scheduler Version: ✔ 3.7.1
Log Directory Writable: ✔
WP Version: 6.4.3
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: TasteWP-S3 Official/3.0.0
PHP Version: 8.0.30
PHP Post Max Size: 128 MB
PHP Time Limit: 100
PHP Max Input Vars: 5000
cURL Version: 7.68.0
OpenSSL/1.1.1f

SUHOSIN Installed: –
MySQL Version: 8.0.32-0ubuntu0.20.04.2
Max Upload Size: 60 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 8.6.1
WC Database Prefix: wp_
Total Database Size: 4.87MB
Database Data Size: 3.18MB
Database Index Size: 1.69MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.08MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 2.03MB + Index: 0.06MB + Engine InnoDB
wp_postmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_posts: Data: 0.06MB + Index: 0.06MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.05MB + Index: 0.02MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.06MB + Index: 0.00MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_addresses: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_operational_data: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_orders: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_wc_orders_meta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wpfm_backup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB

### Post Type Counts ###

attachment: 1
page: 7
post: 2
product: 4
product_variation: 2
revision: 4
shop_order_placehold: 1

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ❌Error messages should not be shown to visitors.

### Active Plugins (1) ###

WooCommerce: by Automattic – 8.6.1

### Dropin Plugins (1) ###

maintenance.php: maintenance.php

### Settings ###

API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to Woo.com: –
Enforce Approved Product Download Directories: ✔
HPOS feature screen enabled: ✔
HPOS feature enabled: ✔
Order datastore: Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore
HPOS data sync enabled: –

### WC Pages ###

Shop base: #6 - /shop/
Cart: #7 - /cart/
Checkout: #8 - /checkout/
My account: #9 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: GeneratePress
Version: 3.4.0
Author URL: https://tomusborne.com
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Admin ###

Enabled Features: activity-panels
analytics
product-block-editor
coupons
core-profiler
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
product-variation-management
product-virtual-downloadable
product-external-affiliate
product-grouped
product-linked
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page

Disabled Features: customize-store
minified-js
new-product-management-experience
product-pre-publish-modal
settings
async-product-editor-category-field

Daily Cron: ✔ Next scheduled: 2024-03-01 16:07:41 +00:00
Options: ✔
Notes: 67
Onboarding: skipped

### Action Scheduler ###

Complete: 14
Oldest: 2024-02-28 16:08:52 +0000
Newest: 2024-02-29 18:17:03 +0000

Failed: 1
Oldest: 2024-02-29 14:40:31 +0000
Newest: 2024-02-29 14:40:31 +0000

Pending: 1
Oldest: 2024-03-01 18:17:03 +0000
Newest: 2024-03-01 18:17:03 +0000

### Status report information ###

Generated at: 2024-03-01 14:58:37 +00:00
`

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woo.com/products/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by ninetyninew**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes #45245 

Note: We originally committed a filter hook for the hardcoded 10 minute reserve stock on block checkout entry, however due to reviews since then we have opted for a filter in `Automattic\WooCommerce\Checkout\Helpers` in the `ReserveStock::reserve_stock_for_order()`.

The proposed newer filter was then reviewed and recommended we amend to a minutes based filter, and hence this is now a `woocommerce_order_hold_stock_minutes` filter hook.

Testing instructions below updated.

**Snippet for testing:**

```
function test_woocommerce_order_hold_stock_minutes( $minutes, $order ) {

	return 0;

}
add_filter( 'woocommerce_order_hold_stock_minutes', 'test_woocommerce_order_hold_stock_minutes', 10, 2 );
```

**Test based on WooCommerce hold stock minutes setting:**

1. First determine that stock is being reserved without the snippet
2. Ensure WooCommerce > Settings > Inventory > Manage stock enabled and set to hold stock for 60 minutes
3. Create a simple product with manage stock on and stock quantity 10
4. Add the product to block cart with quantity 1
5. Place order
6. View order
7. In order notes and/or by checking wp_wc_reserve_stock database table see that stock reserved for 60 minutes (it may also show an earlier 10 minutes note due to the checkout entry draft hardcoded 10 minute stock reserve)
8. Add the snippet
9. Repeat the above steps, this time there shouldn't be any stock reservation (you can also test with a minute level by returning one in the snippet instead of 0)

**Test based on block checkout entry hold stock minutes (this uses a hardcoded 10 minutes not the hold stock setting):**

1. First determine that stock is being reserved for checkout drafts without the snippet
2. Ensure WooCommerce > Settings > Inventory > Manage stock enabled and set to hold stock for 60 minutes (this gets ignored for this test due to the hardcoded 10 minutes)
3. Create a simple product with manage stock on and stock quantity 10
4. Add the product to block cart with quantity 1
5. Go to the checkout page but don't place the order (generates a draft order)
6. View the draft order
7. In order notes and/or by checking wp_wc_reserve_stock database table see that stock reserved for 10 minutes
8. Add the snippet
9. Ensure you empty cart and trash then delete the draft order just to ensure you are back to a clean slate, repeat the above steps, this time there shouldn't be any stock reservation (you can also test with a minute level by returning one in the snippet instead of 0)

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

Add `woocommerce_order_hold_stock_minutes` filter hook to allow the number of minutes stock in an order should be reserved for to be filtered.

#### Comment 

</details>

### CODE REVIEW

**Review by opr — COMMENTED** [review]
Hi @ninetyninew thanks for contributing to WooCommerce. I have some quick feedback on this.

We should ensure the new filter is documented ([similar to how it is done here](https://github.com/woocommerce/woocommerce/blob/1ddd904b260b1d53b07fa97aae5374dc6b683229/plugins/woocommerce/src/Blocks/Domain/Services/CheckoutFields.php#L608-L615)) and maybe a better name would be `woocommerce_draft_order_hold_stock_minutes`.

You could also move it to a separate variable to make it clearer what the documentation refers to.

Finally, the changelog entry on the PR body is incorrectly formatted, but I'll update it for you so it auto-generates one. I'd also change it to an `Add` rather than a tweak since we're adding new functionality here.

I'd like to hear @mikejolley's input on why 10 minutes was chosen, and why it's not editable. It was added in this PR: https://github.com/woocommerce/woocommerce-blocks/pull/5546

@ninetyninew could you also share your use-case on this PR as to why you want to filter this?

**Comment by ninetyninew** [comment]
@opr I've hopefully made those changes and committed them as required.

My use case for this is for when you have a layer of custom stock management on top of WooCommerce's standard stock functionality, it is useful to be able to stop the reservation by suppling zero minutes, previously you could turn off the stock reservation with woocommerce_hold_stock_minutes, but now with Cart/Checkout blocks this isn't possible due to the forced 10 minutes, I have read @mikejolley's ticket on why it is 10 minutes, but I think you should be allowed to supply your own value for this (including 0):

_"we put a hold on stock for all items in the cart to prevent things going out of stock while the customer enters their details"_

If you've set the proposed new filter to 0 and it does go out of stock in that scenario, what would happen?

**Comment by mikejolley** [comment]
> I'd like to hear @mikejolley's input on why 10 minutes was chosen, and why it's not editable. It was added in this PR: https://github.com/woocommerce/woocommerce-blocks/pull/5546

@opr just seemed like a sensible number to allow customer to idle on checkout before invalidating everything. The time gets updated whenever there is an API push during field entry.

If this was somehow set too low, the checkout would error when placing an order if something went out of stock. There would be race conditions.

**Comment by opr** [comment]
Thanks for the info Mike, makes sense.

> If this was somehow set too low, the checkout would error when placing an order if something went out of stock. There would be race conditions.

If we communicate this in the hook docks I think it's OK then.

**Comment by ninetyninew** [comment]
@opr Let me know if anything further needs amending here, I saw yesterday the workflows flagged a lint error which I believe I have rectified in the latest commit.

File: plugins/woocommerce/src/StoreApi/Routes/V1/Checkout.php [hunk 1]
```diff
@@ -391,9 +391,23 @@ private function create_or_update_draft_order( \WP_REST_Request $request ) {
...
+			/**
+			 * Filters the hold stock duration in minutes for draft orders on checkout entry.
+			 *
+			 * This hook filters the duration in minutes that stock is held for draft orders on checkout entry, it allows a third party to amend the duration to increase/reduce the time the stock is held for draft orders on checkout entry.
+			 *
+			 * @since 8.8.0
+			 *
+			 * @param integer $minutes Minutes to hold stock for draft orders on checkout entry.
+			 */
+			$draft_order_hold_stock_minutes = (int) apply_filters( 'woocommerce_draft_order_hold_stock_minutes', 10 );
```
Comment (opr) [inline plugins/woocommerce/src/StoreApi/Routes/V1/Checkout.php]: Maybe we should check that the result is: not empty, is an integer (or can be cast to one) and is not negative, if any of these are true it should default back to 10. Do you agree?

Comment (ninetyninew) [reply plugins/woocommerce/src/StoreApi/Routes/V1/Checkout.php]: @opr I think casting to a integer and a check for negative is fine, however the !empty would stop being able to filter to 0, which for our particular scenario (to stop stock reservation) is why we need this filter.

Comment (opr) [reply plugins/woocommerce/src/StoreApi/Routes/V1/Checkout.php]: Yep, you're right 👍🏼

Comment (ninetyninew) [reply plugins/woocommerce/src/StoreApi/Routes/V1/Checkout.php]: @opr The int casting is already there, i've just amended to include the negative check in the latest commit if you'd like to check over it.

**Comment by opr** [comment]
@ninetyninew thanks for your patience so far, after more testing I have found that if the filter returns 0, then the value is defaulted to 60 in `reserve_stock_for_order` due to this check: https://github.com/woocommerce/woocommerce/blame/8eb8a09c8881bf12a471fd5d650d539308914ae2/plugins/woocommerce/src/Checkout/Helpers/ReserveStock.php#L69

therefore, to disable stock reservation on draft orders, you need to use two filters: `woocommerce_hold_stock_minutes` and `woocommerce_draft_order_hold_stock_minutes`.

Looping in @woocommerce/proton to get your thoughts on this. I personally don't think this is the greatest DX, especially if you _only_ want to set draft order stock reservation time to 0 leaving regular order's reservation time alone.

Instead of this, do you think a better approach would be adding a filter to `reserve_stock_for_order` that allows extensions to decide if the stock should be reserved or not. [It could be added as a condition here](https://github.com/woocommerce/woocommerce/blob/8eb8a09c8881bf12a471fd5d650d539308914ae2/plugins/woocommerce/src/Checkout/Helpers/ReserveStock.php#L71). The filter would receive the `$order` so it could determine the status in the callback.

**Comment by ninetyninew** [comment]
@opr Makes sense, and less complicated. Happy to do that, would I do that here or does it require creation of a new PR? Also what would the suggested filter name be? Maybe `woocommerce_reserve_stock_for_order`?

**Review by opr — APPROVED** [review]
Thanks for working on this @ninetyninew - it's working fine for me now and the stock is not reserved. Please can you close and reopen the PR to get CI running?

**Comment by Konamiman** [comment]
I have a suggestion for an improvement that would make the filter more flexible: what if instead of accepting an returning a boolean, it accepts the value of `$minutes` (from either the function argument or the `woocommerce_hold_stock_minutes` option) and returns a new minutes value? If it returns 0 it means that no stock reservation should be made. This implies that the name of the hook would have to be changed, for example to `woocommerce_order_hold_stock_minutes`.

Thus the begninning of the function body, not including the filter docblock, would look like this (I haven't tested it):
 
```php
$minutes = $minutes ? $minutes : (int) get_option( 'woocommerce_hold_stock_minutes', 60 );
$minutes = apply_filters( 'woocommerce_order_hold_stock_minutes', $minutes, $order );
if ( ! $minutes || ! $this->is_enabled() ) {
	return;
}
```

**Comment by ninetyninew** [comment]
@Konamiman @opr  I find this a little harder to understand than our proposed filter, due to the already existing woocommerce_hold_stock_minutes setting, could that new filter lead to confusion with what woocommerce_hold_stock_minutes is doing maybe? Whereas the woocommerce_reserve_stock_for_order seems quite self explanatory: there is a hold stock minutes setting, but you can exclude an order from it.

Not quite sure on this, but I think that it may also cause a potential issue, as the checkout draft `$minutes` are hardcoded to 10 before it gets here, the filter you are proposing would allow that 10 minute period to be changed to something lower and result in the race conditions @mikejolley mentioned earlier.

**Comment by opr** [comment]
@ninetyninew thanks for sharing your thoughts! I think @Konamiman's suggestion is great and makes a lot of sense, but I also agree that naming is tricky. I think we should go with the new suggestion but with a clearer/easier name.

> the filter you are proposing would allow that 10 minute period to be changed to something lower and result in the race conditions @mikejolley mentioned earlier.

If the stock-hold is set to 0, it's effectively the same as returning `false` in your current filter. Values between 1-10 would not result in race conditions.

**Comment by ninetyninew** [comment]
Thanks for the feedback, I have now made the changes as recommended, committed them and updated the testing instructions. Can you please re-review @opr @Konamiman 

Note: As a $minutes based hook, I can't think of better name than this, when writing the previous reply I was mistakenly forgetting hat woocommerce_hold_stock_minutes isn't a filter in itself and is a setting, so with this new proposed filter being the only filter hook related to what it's doing (and woocommerce_hold_stock_minutes being a setting), I think the name works.

**Comment by Konamiman** [comment]
@ninetyninew It looks good now, thanks. Just one additional small nitpick: it would be good to explicitly state in the docblock of the filter that a value of zero minutes will disable the stock reservation for the order entirely (that was the original goal of the change after all).

**Comment by ninetyninew** [comment]
@Konamiman No problem, I have now made and committed the doc block change.

**Review by Konamiman — APPROVED** [review]

**Comment by ninetyninew** [comment]
@Konamiman Thanks! I noticed 1 thing failed here, do I need to do anything? I haven't raised a PR for a long time so unfamiliar with the latest processes.

**Comment by Konamiman** [comment]
@ninetyninew Don't worry, it's Github trying to create a build from the PR branch but that sometimes fails and it's not related to the code changes introduced.

**Pull request merged**